compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/inference/PostponedArgumentsAnalyzer.kt

```diff
@@ -164,21 +164,30 @@ class PostponedArgumentsAnalyzer(
 
         val checkerSink: CheckerSink = CheckerSinkImpl(candidate)
 
+        val lastExpression = lambda.atom.body?.statements?.lastOrNull() as? FirExpression
         var hasExpressionInReturnArguments = false
+        // Don't constraint for return expressions of lambda if it has Unit return type.
         val lambdaReturnType = lambda.returnType.let(substitute).takeUnless { it.isUnit }
         returnArguments.forEach {
             if (it !is FirExpression) return@forEach
             hasExpressionInReturnArguments = true
-            candidate.resolveArgumentExpression(
-                c.getBuilder(),
-                it,
-                lambdaReturnType,
-                lambda.atom.returnTypeRef, // TODO: proper ref
-                checkerSink,
-                context = resolutionContext,
-                isReceiver = false,
-                isDispatch = false
-            )
+            // If it is the last expression, and the resolved type (via expected type) is Unit, that expression will be coerced to Unit.
+            // If the last expression is of Unit type, of course it's not coercion-to-Unit case.
+            val lastExpressionCoercedToUnit =
+                it == lastExpression && lambda.atom.returnTypeRef.coneType.isUnit && !it.typeRef.coneType.isUnit
```

Comment:
Indeed, lambda return type is not inferred _yet_ here, and we don't want to constrain it if the last expression is going to be coerced-to-Unit (as we did for the inferred lambda return type if that inferred type is `Unit`). Otherwise, the last expression is revisited, and in the original example, `null` of `Nothing?` type will force the lambda return type and the block type again to be `Nothing?`.

Regarding `lambdaReturnType` variable, yeah, it's still type variable `T` in this example, since it's not fixed yet. But, again, we don't want to constrain it as `Nothing?`, since we will do coercion-to-Unit. To be clearer, I changed the call site to pass the explicit expected return type.

FYI, here is the counterpart in old FE: https://github.com/JetBrains/kotlin/blob/master/compiler/frontend/src/org/jetbrains/kotlin/resolve/calls/tower/KotlinResolutionCallbacksImpl.kt#L230 where `expectedReturnType` is indeed `expectedTypeForReturnArguments` (https://github.com/JetBrains/kotlin/blob/master/compiler/resolution/src/org/jetbrains/kotlin/resolve/calls/components/PostponedArgumentsAnalyzer.kt#L130), which matches with the change I add here.